internal/storage/chunk_uploader.go

```diff
@@ -0,0 +1,202 @@
+	state UploaderState
+}
+
+// NewChunkUploader creates a new instance of chunkUploader,
+// for the given inputs.
+func NewChunkUploader(ctx context.Context, obj *storage.ObjectHandle, req *gcs.CreateObjectRequest, writeChunkSize int, progressFunc func(int64)) (gcs.ChunkUploader, error) {
+	if ctx == nil {
+		return nil, fmt.Errorf("ctx is nil")
+	}
+	if obj == nil || req == nil {
+		return nil, fmt.Errorf("nil obj or req")
+	}
+
+	if obj.ObjectName() != req.Name {
+		return nil, fmt.Errorf("names of passed object-handle and createObjectRequest.Name don't match: object-handle=%s createObjectRequest.Name=%s", obj.ObjectName(), req.Name)
+	}
+
+	if req.GenerationPrecondition != nil && *req.GenerationPrecondition != 0 {
+		return nil, fmt.Errorf("request received for pre-existing object %s, supported only for new objects", req.Name)
+	}
+
+	if writeChunkSize <= 0 {
+		return nil, fmt.Errorf("chunkSize = 0")
+	}
+
+	// Raw initialization.
+	uploader := chunkUploader{}
+
+	// Store references to necessary parameters.
+	uploader.objectName = obj.ObjectName()
+	uploader.userProgressFunc = progressFunc
+
+	// Create a NewWriter with the requested attributes, using Go Storage Client.
+	// NewWriter never returns nil, so no nil-check is needed on it.
+	wc := obj.NewWriter(ctx)
+	wc = storageutil.SetAttrsInWriter(wc, req)
+	wc.ChunkSize = writeChunkSize
+	wc.ProgressFunc = func(n int64) {
+		uploader.totalWriteSucceededSoFar = n
+		logger.Debugf("%d bytes copied so far for object/file %s. chunk-size = %d", n, req.Name, wc.ChunkSize)
+
+		if uploader.userProgressFunc != nil {
+			uploader.userProgressFunc(n)
+		}
+	}
+
+	uploader.wc = wc
+	uploader.state = Initialized
+
+	return &uploader, nil
+}
+
+func (uploader *chunkUploader) chunkSize() (int, error) {
+	if uploader.wc == nil {
+		return 0, fmt.Errorf("writer not initialized")
+	}
+	return uploader.wc.ChunkSize, nil
+}
+
+// BytesUploadedSoFar returns the total number of bytes successfully uploaded
+// so far using this uploader.
+// This waits (using mutex) until the completion of
+// in-progress calls to BytesUploadedSoFar/UploadChunkAsync/Close
+// invoked from other threads/go-routines.
+func (uploader *chunkUploader) BytesUploadedSoFar() int64 {
+	uploader.mu.Lock()
+	defer uploader.mu.Unlock()
+
+	return uploader.totalWriteSucceededSoFar
+}
+
+func (uploader *chunkUploader) readyToWrite() bool {
+	switch uploader.state {
+	case Initialized, Waiting, Writing:
+		return true
+	default:
+		return false
+	}
+}
+
+func (uploader *chunkUploader) readyToClose() bool {
+	switch uploader.state {
+	case Initialized, Waiting, Writing, WriteError:
+		return true
+	default:
+		return false
+	}
+}
+
+// UploadChunkAsync uploads the passed content to GCS.
+// This waits (using mutex) until the completion of
+// in-progress calls to BytesUploadedSoFar/UploadChunkAsync/Close
+// invoked from other threads/go-routines.
+func (uploader *chunkUploader) UploadChunkAsync(contents io.Reader) error {
```

Comment:
> This method has no logic related to async upload

The word async in function name indicates that this function doesn't block on the upload to complete. This is a common convention.

> I am a little concerned about retries and timeouts for this call since it doesn't take context. Context aware reader is one option but I was wondering if go client has something related?

Let's take this separately in another PR. This needs discussion and possibly interface change.